components/icu4x_ecma402/src/pluralrules.rs

```diff
@@ -0,0 +1,163 @@
+//! TODO: License blurb.
+
+//! Implements the traits found in [ecma402_traits::pluralrules].
+
+use ecma402_traits;
+use icu_pluralrules as ipr;
+use icu_pluralrules::PluralRulesError;
+
+pub(crate) mod internal {
+    use ecma402_traits::pluralrules::options::Type;
+    use ecma402_traits::pluralrules::Options;
+    use icu_pluralrules::{PluralCategory, PluralOperands, PluralRuleType};
+
+    /// Converts the trait style option to an equivalent ICU4X type.
+    pub fn to_icu4x_type(style: &Type) -> PluralRuleType {
+        match style {
+            Type::Ordinal => PluralRuleType::Ordinal,
+            Type::Cardinal => PluralRuleType::Cardinal,
+        }
+    }
+
+    /// Converts the number to format into the operands representation.
+    pub fn to_icu4x_operands(n: f64, _opts: Options) -> PluralOperands {
+        // TODO: This does not convert the display Options correctly but
+        // probably should.
+        let nstr = format!("{}", n);
+        nstr.parse().unwrap()
+    }
+
+    /// Expresses the [PluralCategory] as a `str`.
+    pub fn as_str(c: PluralCategory) -> &'static str {
+        match c {
+            PluralCategory::Few => "few",
+            PluralCategory::Many => "many",
+            PluralCategory::One => "one",
+            PluralCategory::Other => "other",
+            PluralCategory::Two => "two",
+            PluralCategory::Zero => "zero",
+        }
+    }
+}
+
+pub struct PluralRules {
+    opts: ecma402_traits::pluralrules::Options,
+    rep: ipr::PluralRules,
+}
+
+impl ecma402_traits::pluralrules::PluralRules for PluralRules {
+    type Error = PluralRulesError;
+
+    fn try_new<L>(l: L, opts: ecma402_traits::pluralrules::Options) -> Result<Self, Self::Error>
+    where
+        L: ecma402_traits::Locale,
+        Self: Sized,
+    {
+        // Which data provider should be the default here?
+        let dp = icu_data_provider::InvariantDataProvider;
```

Comment:
Ah, so here you will generate a new DataProvider for each instance of `PluralRules`... That's unfortunate I think.

I'm wondering if it would be possible to have some Trait object that would implement `get_provider` and it would be some `IntlEnvironment` which you have a singleton of, and each constructor just uses the provider from it.

Then you'd likely do:

```rust
let intl = IntlEnvironment::new(provider);

let pr = intl.try_new_plural_rules(lang, options);

let dtf = intl.try_new_date_time_format(lang, options);
```

etc. I'm not in love with this UX, but it does feel right from the ergonomics perspective.

Maybe there's a better way to associate a single `DataProvider` with all constructors?

Maybe something like:

```rust
let pr = intl.PluralRules::try_new(lang, options)?;
```

@ManishEarth - any thoughts on this?

Reply:
Frankly I would rather remove the requirement that the constructor be part of the trait rather than including yet another necessary trait to implement.  IMHO the coupling there is quite tight but doesn't justify the gains.

If anything, we could, ostensibly, add a constructor arg to pass arbitrary data into the constructor if we so wish.

Reply:
I'd opt to leave the provision as is for the time being, and retrofit a "default provider".  

This won't change the API observed by the end user.